### PR TITLE
WIP: IntoIterator for Box<[T]> + method dispatch mitigation for editions < 2024

### PR DESCRIPTION
**Pull request by clarfonthey**
ACP: rust-lang/libs-team#263
References #59878

Recommendation per ACP: this should receive a crater run to gauge impact. If there's no impact, it can be merged as-is, but otherwise it will need a similar edition-based workaround to the array implementation.

In addition to what was proposed by the ACP, this also adds `IntoIterator for &Box<[T]>` and `IntoIterator for &mut Box<[T]>` to ensure that those work as expected. I also already had to change at least one line in the compiler to account for this change, which isn't a good sign toward whether edition-specific mitigations may be needed, but we'll see.

### CODE REVIEW

**Comment by clarfonthey** [comment]
It appears that Clippy is explicitly testing for this, so, I'll have to take a look at that. But it should at least be ready for a Crater run if we're looking to just judge the impact of the initial impl.

File: library/alloc/src/boxed.rs [hunk 1]
```diff
@@ -2036,6 +2038,42 @@ impl<I> FromIterator<I> for Box<[I]> {
...
+impl<I, A: Allocator> IntoIterator for Box<[I], A> {
+    type IntoIter = vec::IntoIter<I, A>;
+    type Item = I;
+    fn into_iter(self) -> vec::IntoIter<I, A> {
+        self.into_vec().into_iter()
+    }
+}
+
+#[stable(feature = "boxed_slice_into_iter", since = "CURRENT_RUSTC_VERSION")]
+impl<'a, I, A: Allocator> !Iterator for &'a Box<[I], A> {}
```
Comment (WaffleLapkin) [inline library/alloc/src/boxed.rs:2054]: why is this needed? is the `!Iterator` impl for `[_]` not enough?

Comment (clarfonthey) [reply library/alloc/src/boxed.rs:2054]: Unfortunately not. :(

**Comment by Mark-Simulacrum** [comment]
@craterbot check

**Comment by clarfonthey** [comment]
I'm not 100% clear on the results (it looks like more than 9 root crates are actually failing), but it does seem to me like some sort of edition mechanism would be good to include. Is that the kind of thing that should have an RFC, or would it be reasonable to say that the ACP is enough?

**Comment by clarfonthey** [comment]
After taking a look at what was done for 2021 edition, it appears that there wasn't any particular RFC for it, just, a PR was made and that was the solution. So, I'm going to follow suit and start work on the override, hopefully not having to tread too deeply in compiler internals.

After a quick search, it shouldn't be too difficult to modify the `rustc_skip_array_during_method_dispatch` attribute to work with boxed slices, and to create a `boxed_slice_into_iter` edition lint in addition to the `array_into_iter` edition lint.

The only major questions I have here are:

* It appears that the method dispatch code is replicated inside the `rust-analyzer` code base, and I'm going to assume I shouldn't touch that for now, and `rust-analyzer` can be fixed later.
* The edition guides exist inside the repo but the 2024 edition guide hasn't been made yet, so, I'm not sure if I should write up something for this similar to that for the 2021 edition on arrays, or if I should hold off for now.
* I don't think that this should need an FCP assuming that there are tests that verify that the behaviour is the same on the 2021 edition, although I'm assuming that once everything is done, we'll want to do a crater run again to verify that.

**Comment by clarfonthey** [comment]
Will try and return to this PR this week. I would like to see this in the 2024 edition.

**Comment by clarfonthey** [comment]
(did a very crude rebase, and the previous code wasn't even working, so. gonna poke at this more later)

File: library/alloc/src/boxed.rs [hunk 2]
```diff
@@ -2073,6 +2114,45 @@ impl<I> FromIterator<I> for Box<[I]> {
...
 
+#[stable(feature = "boxed_slice_into_iter", since = "CURRENT_RUSTC_VERSION")]
+impl<I, A: Allocator> !Iterator for Box<[I], A> {}
+
+// Note: the `#[rustc_skip_during_method_dispatch(boxed_slice)]` on `trait IntoIterator`
+// hides this implementation from explicit `.into_iter()` calls on editions < 2024,
+// so those calls will still resolve to the slice implementation, by reference.
+#[stable(feature = "boxed_slice_into_iter", since = "CURRENT_RUSTC_VERSION")]
+impl<I, A: Allocator> IntoIterator for Box<[I], A> {
+    type IntoIter = vec::IntoIter<I, A>;
```
Comment (svix-jplatte) [inline library/alloc/src/boxed.rs:2125]: I wonder, should this really be publically the same type? (can't be changed later)

Comment (WaffleLapkin) [reply library/alloc/src/boxed.rs:2125]: It's hard to imagine why we would want something different.

`vec::IntoIter` is
- ptr + cap + alloc (needed to free allocation)
- ptr to first not yielded element
- ptr one past the last not yielded element

You need all the same things for a box. Since the size of initialized elements changes when iterating, there is no difference with a vec -- you need to save the allocation and you need to save the initialized range in it.

Comment (clarfonthey) [reply library/alloc/src/boxed.rs:2125]: Also worth pointing out that *not* making them the same type is also something we can't revert after the fact, so, it feels like we should just pick the logical option here and make them the same.

**Comment by cuviper** [comment]
Should we also do this for `Box<[T; N]>`? That could *possibly* deserve a separate iterator type, since the capacity is hard-coded in `N`, but I think it would also be fine to just use `vec::IntoIter` as well.

**Comment by clarfonthey** [comment]
I feel like `Box<[T; N]>` is actually a bit more controversial for the type since it might be valuable to keep the `N` as part of the iterator type, but we also don't want to move it *off* the heap when deconstructing.

That one I feel like we could potentially copy `array::IntoIter` and make a proper version, which then makes me question whether we should just do the same for regular boxes too just to be consistent.

**Comment by m-ou-se** [comment]
We discussed this in today's libs-api meeting. We'd be happy to see this in the next edition.

We didn't think it is worth making separate IntoIter types, and would prefer simply reusing Vec's IntoIter (for both box-of-slice and box-of-array). Keeping the `N` part of the type doesn't seem to have a lot of benefits, as it only represents an upper bound, which has very limited use.

**Comment by the8472** [comment]
Marking as T-lang since it affects method dispatch.

**Comment by clarfonthey** [comment]
I also have been meaning to get back to this but kept running into unrelated life issues. Will hopefully be able to do a proper dive soon-- the deadline to get an initial implementation in two weeks seems pretty doable; my plan, as demonstrated by the existing mess, is to just piggyback off the 2021 edition implementation.

File: library/alloc/src/boxed.rs (hunk 1 above)
Comment (compiler-errors) [reply library/alloc/src/boxed.rs:2133]: A comment here for why they're needed would be very 🆒

**Comment by clarfonthey** [comment]
Okay, I thought about it and I simply cannot justify putting in this much work with this little help given the other things I have going on in life. Would love to see this in edition 2024, and I'm sure there are folks knowledgeable in this code who could pick up the slack, but if it's pushed to a future edition, that's fine too.

I don't want this enough to abandon other priorities at the moment, especially as I'm unemployed and this effort means nothing to any job I apply to. Apologies but also, not sorry for my decision. Will close this and folks can decide what to do.

**Comment by compiler-errors** [comment]
I'm sorry to hear that @clarfonthey. I'll make sure to pick this up for the edition, and make sure to keep the commits with you as the author, unless you don't want to -- ping me on Zulip if so. Thanks for all the effort you put into this up til now!
